### PR TITLE
Feat/apply incremental schema changes

### PR DESCRIPTION
**Pull request by canbekley**
## Summary
implementation of schema changes for incremental models are currently missing. detecting schema changes during materialization would fallback to the legacy incremental strategy. this is not user-friendly and not feasible for very large models. schema changes for distributed_incremental materialization were not supported at all.

this pr contains
- support for applying schema changes for `incremental` and `distributed_incremental` materializations
- implementation of the `on_schema_change='sync_all_column'` strategy as per official [dbt specs](https://docs.getdbt.com/docs/build/incremental-models#what-if-the-columns-of-my-incremental-model-change)
- fix of an issue in distributed_incremental materialzation and `on_schema_change='ignore'`, where new columns in the target relation would break the dbt run during creation of the `__dbt_new_data_` tables
- fix of an issue in distributed_incremental materialzation, where column order changes would break the dbt run (or insert data to the wrong columns)

## Checklist
Delete items not relevant to your PR:
- [x] Unit and integration tests covering the common scenarios were added
- [x] A human-readable description of the changes was provided to include in CHANGELOG
- [ ] For significant changes, documentation in https://github.com/ClickHouse/clickhouse-docs was updated with further explanations or tutorials

### CODE REVIEW

File: tests/integration/adapter/incremental/test_schema_change.py [hunk 1]
```diff
@@ -4,68 +4,183 @@
...
+        result = project.run_sql(f"select * from {model} order by col_1", fetch="all")
         assert len(result) == 3
         assert result[0][1] == 1
-        run_dbt(["--debug", "run", "--select", "schema_change_append"])
-        result = project.run_sql("select * from schema_change_append order by col_1", fetch="all")
+        run_dbt(["--debug", "run", "--select", model])
+        result = project.run_sql(f"select * from {model} order by col_1", fetch="all")
         assert result[0][2] == 0
         assert result[3][2] == 5
+
```
Comment (BentsiLeviav) [inline tests/integration/adapter/incremental/test_schema_change.py:76]: Would you mind also adding a test to check the drop columns logic?

File: dbt/adapters/clickhouse/impl.py [hunk 2]
```diff
@@ -193,35 +193,33 @@ def calculate_incremental_strategy(self, strategy: str) -> str:
     @available.parse_none
     def check_incremental_schema_changes(
         self, on_schema_change, existing, target_sql
-    ) -> List[ClickHouseColumn]:
-        if on_schema_change not in ('fail', 'ignore', 'append_new_columns'):
-            raise DbtRuntimeError(
-                "Only `fail`, `ignore`, and `append_new_columns` supported for `on_schema_change`"
-            )
+    ) -> ClickHouseColumnChanges:
```
[the remark this answers is not in the document]
Comment (canbekley) [reply dbt/adapters/clickhouse/impl.py:196]: of course 👍  I added them back

File: dbt/include/clickhouse/macros/materializations/incremental/incremental.sql [hunk 3]
```diff
@@ -232,5 +231,4 @@
         insert into {{ existing_relation }} select {{ dest_cols_csv }} from {{ inserting_relation }} {{ adapter.get_model_query_settings(model) }}
     {% endcall %}
     {% do adapter.drop_relation(new_data_relation) %}
-    {{ drop_relation_if_exists(distributed_new_data_relation) }}
```
[the remark this answers is not in the document]
Comment (canbekley) [reply dbt/include/clickhouse/macros/materializations/incremental/incremental.sql]: I did move this up by mistake. I just saw that the distributed_new_data_relation is actually dropped in the create_distributed_local_table. I reverted it

File: tests/integration/adapter/incremental/test_schema_change.py (hunk 1 above)
Comment (canbekley) [reply tests/integration/adapter/incremental/test_schema_change.py:77]: added assertions for the dropped column 👍

Comment (canbekley) [reply tests/integration/adapter/incremental/test_schema_change.py:77]: the drop column logic is now tested in TestComplexSchemaChange:test_sync. we are just asserting the resulting column count at the moment. I could extend it to check for existing column names as well

File: dbt/include/clickhouse/macros/materializations/incremental/incremental.sql (hunk 3 above)
[the remark this answers is not in the document]
Comment (canbekley) [reply dbt/include/clickhouse/macros/materializations/incremental/incremental.sql]: the initial changes I made here were reverted

**Comment by BentsiLeviav** [comment]
Github changed its `docker-compose` command so all the tests failed. I adjusted it in #334 and merged it to main.
@canbekley Could you please merge `main` to your branch?

**Pull request merged**